### PR TITLE
Add Wav2vec2-finetuning notebook

### PR DESCRIPTION
**Pull request by vasudevgupta7**
This PR will add a notebook showing Wav2Vec2 SavedModel fine-tuning on LibriSpeech dataset.

Link to notebook: https://colab.research.google.com/github/vasudevgupta7/hub/blob/wav2vec2-notebook/examples/colab/wav2vec2_saved_model_finetuning.ipynb

@MorganR @sayakpaul

### CODE REVIEW

File: examples/colab/wav2vec2_saved_model_finetuning.ipynb [hunk 1]
```diff
@@ -0,0 +1,2441 @@
+{
```
Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:1]: <p>nit: These pre-trained weights <em>were</em> converted from ...</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/hub/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:1]: <p>nit: with Keras's Functional API.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/hub/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:1]: <p><s>get out-of-box performance</s> -&gt; improve performance further.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/hub/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:1]: <p>build -&gt; built</p><p> </p><p>when model.<strong>call</strong> or model.build are called, so ...</p><p> </p><p>for checking -&gt; to check</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/hub/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:1]: <p>need small data -&gt; a small amount of data.</p><p> </p><p>Could you also mention how to get the whole dataset?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/hub/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:1]: <p>nit: in <em>the</em> LibriSpeech directory. Let's explore these files.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/hub/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:1]: <p>is having -&gt; has</p><p> </p><p>and single <code>.txt</code>  -&gt; and a single <code>.txt</code>  </p><p> </p><p><em>The</em> <code>.txt</code>  file contains text transcriptions ...</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/hub/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:1]: <p>We can load this text data as follows:</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/hub/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:1]: <p>for loading <em>a</em> speech sample from <em>a</em> <code>.flac</code>  file.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/hub/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='30'/>

Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:1]: <p>It would be nice to explain more about what tokenizer and processor are doing</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/hub/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='38'/>

Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    generator = partial(inputs_generator, samples=samples)
</pre><p>I don't think this block is needed; it's pretty clear at this point what's going on. It would be more clear to combine the <code>def inputs_generator</code>  block with the <code>tf.data.Dataset.from_generator...</code>  block below.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/hub/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='42'/>

Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:1]: <p>It's also worth mentioning that this code is loading the whole dataset into memory, which is fine for the small dataset here, but will not work with the full-sized LibriSpeech dataset.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/hub/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='43'/>

Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:1]: <p>I think this can be removed. I expect users to understand shuffle, or read the documentation if they're confused.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/hub/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='45'/>

Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:1]: <p>This is actually quite interesting. Maybe it's beyond the scope of this colab, but I would expect that blindly truncating like this could negatively impact training. For any samples that are truncated, there is no longer a guarantee that the remaining labels match all the remaining audio, so I would think it's better to discard these samples.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/hub/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='48'/>

Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:1]: <p>It would be good to clarify why you're writing a custom train and eval step (eg. there isn't built-in keras functionality for some part of this, etc)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/hub/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='53'/>

Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #4.    pbar = tqdm(range(EPOCHS), total=EPOCHS)
</pre><p><code>tqdm</code> comes from your wav2vec repo, right? It might be good to give it a more clear name if it's going to be relied on externally like this.</p><p> </p><p>Similarly consider renaming <code>pbar</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/hub/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='56'/>

Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    save_dir = "finetuned-wav2vec2"
</pre><p>It might be worth mentioning that they can set <code>include_optimizer=False</code>  if this is being exported only for inference. This keeps the model a bit smaller.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/hub/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='73'/>

**Comment by gustheman** [comment]
The colab looks good:

some points:
- you mention that the KerasLayer maker the model opaque, does that affect this colab? opaque in which sense?
- 
- why not use the librispeech from tf datasets? (https://www.tensorflow.org/datasets/catalog/librispeech)
- I don't know if it's needed to query the GPU at the begin of the colab. Our samples usually don't do that. I'd add a comment telling the user to make sure they are using a GPU
- it's also not clear to me why doing the custom loop instead of the fit method directly. I guess some more clarification would help the users
- the use of the metric "wer" could be explained in more depth. maybe a link to the documentation directly (I couldn't find it)
- still on the metric, is it used in the final model (Trainer class)?

**Comment by sayakpaul** [comment]
> you mention that the KerasLayer maker the model opaque, does that affect this colab? opaque in which sense?

If the model is loaded in `KerasLayer` you lose access to the inner layers of the model which can be of importance to researchers and practitioners. 

Often there's a need to gradually fine-tune a couple of layers to squeeze every ounce of performance. This use-case is likely not possible if the model is loaded in a `KerasLayer`.

**Comment by gustheman** [comment]
> > you mention that the KerasLayer maker the model opaque, does that affect this colab? opaque in which sense?
> 
> If the model is loaded in `KerasLayer` you lose access to the inner layers of the model which can be of importance to researchers and practitioners.
> 
> Often there's a need to gradually fine-tune a couple of layers to squeeze every ounce of performance. This use-case is likely not possible if the model is loaded in a `KerasLayer`.

but does it affect this colab specifically? otherwise, as a developer I can think: why don't I use the other method instead right from the start?

**Comment by sayakpaul** [comment]
It does not. My answer was to address - "opaque in which sense?".

File: examples/colab/wav2vec2_saved_model_finetuning.ipynb (hunk 1 above)
Comment (vasudevgupta7) [reply examples/colab/wav2vec2_saved_model_finetuning.ipynb:1]: tqdm is not coming from my project, rather using [this package](https://tqdm.github.io).

**Comment by vasudevgupta7** [comment]
@gustheman, thank you so much for the feedback. I am really sorry for the late reply.

> * why not use the librispeech from tf datasets? (https://www.tensorflow.org/datasets/catalog/librispeech)

TF dataset is downloading complete dataset (~300 GBs) which is out of scope of Colab notebook. Hence downloading the dataset from official website.

> * it's also not clear to me why doing the custom loop instead of the fit method directly. I guess some more clarification would help the users

I have removed custom loop and relying on fit method (You will see the update in my next commit).

> * still on the metric, is it used in the final model (Trainer class)?

It's hard to implement WER in `tf.keras.Metric` and hence measuring WER explicitly while doing evaluation. I have made `Evaluation` section more clear now.

**Comment by vasudevgupta7** [comment]
> but does it affect this colab specifically? otherwise, as a developer I can think: why don't I use the other method instead right from the start?

I agree with you on that; but `tf.keras.models.load_model(...)` can't load directly from TFHub (we will have to rely on GCS buckets or downloading using wget) and this notebook is meant to show how model can be loaded from TFHub without any extra effort. Hence loading with `hub.KerasLayer`.

**Comment by vasudevgupta7** [comment]
@gustheman, I have worked on all the suggestions. Please review it again (sorry for all the delay from my side).

gently pinging @sayakpaul for keeping updated with the progress.

**Comment by gustheman** [comment]
Thanks Vasudev!

LGTM from my side

**Review by MorganR — APPROVED** [review]

File: examples/README.md [hunk 2]
```diff
@@ -2,9 +2,11 @@
 
 ## Notebooks
 
-#### [`colab/text_classification_with_tf_hub_on_kaggle.ipynb`](colab/text_classification_with_tf_hub_on_kaggle.ipynb)
+### [`colab/`]
```
Comment (MorganR) [inline examples/README.md:5]: @vasudevgupta7 can you fix this? Looks like you accidentally deleted the old first example

**Review by MorganR — CHANGES_REQUESTED** [review]
Mostly LGTM here as well. Please fix my last comment and squash your commits into a single commit.

**Comment by vasudevgupta7** [comment]
@MorganR, done.

**Review by MorganR — APPROVED** [review]

File: examples/colab/wav2vec2_saved_model_finetuning.ipynb [hunk 3]
```diff
@@ -0,0 +1,1946 @@
...
+  "nbformat": 4,
```
Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:2]: Need to add a copyright header, license, and TF buttons, like in https://github.com/tensorflow/hub/blob/master/examples/colab/yamnet.ipynb.

Essentially copy the first three cells from that notebook, but replace "yamnet" in the button links with "wav2vec2_saved_model_finetuning", and update the TF Hub link.

File: examples/colab/wav2vec2_saved_model_finetuning.ipynb [hunk 4]
```diff
@@ -0,0 +1,1946 @@
...
+  "nbformat_minor": 0,
```
Comment (MorganR) [inline examples/colab/wav2vec2_saved_model_finetuning.ipynb:3]: Please save your notebook with the table of contents visible (this should add `"toc_visible": true` to the raw file).

**Review by MorganR — CHANGES_REQUESTED** [review]
A few more changes needed to make internal systems happy.

**Comment by vasudevgupta7** [comment]
@MorganR, internal system would be happy now hopefully :)

**Review by MorganR — APPROVED** [review]
Looks resolved, I'll try to merge it again.

**Review by MorganR — CHANGES_REQUESTED** [review]
Ok, hopefully this is the last change. You need to clear all the outputs from the notebook.

**Comment by vasudevgupta7** [comment]
@MorganR, done.

**Review by MorganR — APPROVED** [review]
LGTM again

**Review by MorganR — CHANGES_REQUESTED** [review]
In the cell where you `pip3` install your repo, can you also add:

```
!sudo apt-get install -y libsndfile1-dev
!pip3 install -q SoundFile
```

I have confirmed that this resolves some final internal issues.

**Comment by vasudevgupta7** [comment]
@MorganR, done. Thanks for the review!!

**Review by MorganR — APPROVED** [review]

File: examples/README.md [hunk 5]
```diff
@@ -2,6 +2,10 @@
 
 ## Notebooks
 
+#### [`colab/wav2vec2_saved_model_finetuning.ipynb`](colab/wav2vec2_saved_model_finetuning.ipynb)
+
+Shows how to fine-tune Wav2Vec2 SavedModel with a language modeling head for Automatic Speech Recognition
```
Comment (MorganR) [inline examples/README.md:7]: Lines need to be 80 characters or less. Please split this into two lines.

**Review by MorganR — CHANGES_REQUESTED** [review]

File: examples/README.md (hunk 5 above)
Comment (vasudevgupta7) [reply examples/README.md]: @MorganR, done.

**Review by MorganR — APPROVED** [review]
